design/design-doc-composition-functions.md

```diff
@@ -175,32 +179,557 @@ metadata:
+    # uses Go Templates (like Helm) to render composed resources.
+    functionRef:
+    - name: go-templates
+    # A Function may optionally accept configuration. The configuration is
+    # passed to the Function when it is called. The configuration block looks
+    # like a nested custom resource - i.e. it has an apiVersion and kind.
+    config:
+      apiVersion: example.org/v1
+      kind: GoTemplate
+      source: Remote
+      remote: git://github.com/example/my-xpostgresql-go-templates
 ```
 
-Under this proposal each function is the entrypoint of an OCI image, though the
-API is designed to support different function implementations (such as webhooks)
-in the future. The updated API would affect only the `Composition` type - no
-changes would be required to the schema of `CompositeResourceDefinitions`, XRs,
-etc.
+The updated API would affect only the `Composition` type - no changes would be
+required to the schema of `CompositeResourceDefinitions`, XRs, etc.
 
-Notably the functions would not be responsible for interacting with the API
-server to create, update, or delete composed resources. Instead, they instruct
-Crossplane which resources should be created, updated, or deleted.
+Notably the Functions would not need to be responsible for interacting with the
+API server to create, update, or delete composed resources. Instead, they
+instruct Crossplane which resources should be created, updated, or deleted.
 
-Under the proposed design functions could also be used for purposes besides
+Under the proposed design Functions could also be used for purposes besides
 rendering composed resources, for example validating the results of the
-`resources` array or earlier functions in the `functions` array. Furthermore, a
-function could also be used to implement 'side effects' such as triggering a
+`resources` array or earlier Functions in the `functions` array. Furthermore, a
+Function could also be used to implement 'side effects' such as triggering a
 replication or backup.
 
-Below is a more detailed example of an entry in the `functions` array.
+Before you can use a Function, you must install it. Installing a Function works
+just like installing a Provider:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: go-templates
+spec:
+  package: xpkg.upbound.io/negz/go-templates:v0.1.0
+```
+
+The Function package's `metadata.name` corresponds to the `functionRef` in the
+previous example Composition. A Composition specifies which Function to run by
+referencing the package's name. Like any other package, the `Function` type will
+have a corresponding `FunctionRevision` type.
+
+### Calling a Function
+
+Despite the name, a 'Function' is actually more like a 'function server'. Under
+this proposal, Functions are long-running processes. When you install one, the
+package manager deploys it using a Kubernetes Deployment - the same way it would
+deploy a Provider.
```

Comment:
What about calling the "function server" `FunctionRunner` and `Function` an actual configuration to be run with a function runner? I mean a function to me is something that takes an input and produce some output, so something like the following:
```yaml
---
apiVersion: pkg.crossplane.io/v1beta1
kind: FunctionRunner
metadata:
  name: go-templates
spec:
  package: xpkg.upbound.io/negz/go-templates:v0.1.0
status:
  # The gRPC endpoint where Crossplane will send RunFunctionRequests.
  endpoint: https://go-templates-9sdfn2
---
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: my-xpostgresql-go-templates
spec:
  functionRef:
    name: go-template
  config:
    apiVersion: example.org/v1
    kind: GoTemplate
    source: Remote
    remote: git://github.com/example/my-xpostgresql-go-templates
---
apiVersion: apiextensions.crossplane.io/v2alpha1
kind: Composition
metadata:
  name: example
spec:
  compositeTypeRef:
    apiVersion: database.example.org/v1
    kind: XPostgreSQLInstance
  functions:
  - functionRef:
      name: my-xpostgresql-go-templates
```
which could be inlined as follows:
```yaml
---
apiVersion: apiextensions.crossplane.io/v2alpha1
kind: Composition
metadata:
  name: example
spec:
  compositeTypeRef:
    apiVersion: database.example.org/v1
    kind: XPostgreSQLInstance
  functions:
  - functionRunnerRef:
      name: go-templates
      config:
        apiVersion: example.org/v1
        kind: GoTemplate
        source: Remote
        remote: git://github.com/example/my-xpostgresql-go-templates
```

Reply:
> I mean, a function to me is something that takes an input and produce some output

I think that's the "trick" here. We're really trying to frame these things (that are in practice more like servers) as "functions". We want folks to feel like they're building, installing support for, and calling "just a function" - not a server or runner. The fact that we make it a server at build time is a bit of an "implementation detail".

So to that end I think I prefer sticking with just "function" rather than introducing a runner type.

I see the (optional) config block in a Composition as kind of like an argument to a function call. In pseudocode Crossplane is making a call like this:

```
run_function(xr, composedResources, config) (xr, composedResources)
```

I think there will be a desire to avoid repeating the same config across multiple function calls, but I'd rather do that by allowing Functions to deliver a Config CRD, which could be instantiated and referenced. This has the added benefit of allowing Functions to deliver fully (OpenAPI) schemafied config types that the API server could validate.

Reply:
@turkenh @jbw976 what do you think here?

Reply:
I think my confusion comes from the fact that I would think to be writing more “runtime” based functions than full-fledged ones, hence my point of separating functions and function runners (a.k.a. runtimes).

Maybe it’s because I’m thinking of something similar to [this](https://docs.konghq.com/hub/kong-inc/serverless-functions/) which I used in the past, but as a user I would expect to be writing more quick “scripts” in Python, Starlark or some other scripting language, rather than really complex “functions” that would require to be wrapped in a grpc server on their own.

But I feel your focus is instead more on full-fledged functions, so it makes sense to consider “runtime” ones as an exception/hack w.r.t. the concept of a function, in which one of the arguments to the function being called is actually custom logic/code, and therefore don’t deserve being first-class citizens of the API.

Reply:
Totally agree with @sttts that it's mainly cosmetics, so definitely not a blocker.

Both approaches have pros and cons, for example, going with FunctionRunners + Functions, how would you define a full-fledged single-purpose function built with the flow you are describing? 
With this model it would have to be defined as a FunctionRunner and should either be referenced directly or through a empty useless Function, but it's not going to run other functions. A solution to this could be to have a `kubernetes` `FunctionRunner` which would take the grpc server as an OCI image built with the flow you are describing through a Function and deploy it as an additional deployment.
Something like this:
```yaml
---
apiVersion: pkg.crossplane.io/v1beta1
kind: FunctionRunner
metadata:
  name: kubernetes
spec:
  package: xpkg.upbound.io/crossplane/kubernetes-xfn-runner:v0.1.0
---
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: my-awesome-func
spec:
  functionRunnerRef:
	name: kubernetes
  config:
	apiVersion: example.org/v1
	kind: Kubernetes
	package: phisco/my-awesome-func:v0.1.0
---
apiVersion: apiextensions.crossplane.io/v2alpha1
kind: Composition
metadata:
  name: example
spec:
  compositeTypeRef:
    apiVersion: database.example.org/v1
    kind: XPostgreSQLInstance
  pipeline:
  - step: compose-xr-using-my-awesome-func
    functionRef:
      name: my-awesome-func
  - step: validate-composed-resources
    functionRef:
      name: cel-validation
```
This would allow decoupling the interface between the Crossplane and `FunctionRunners` from that between the `kubernetes FunctionRunner` and functions built with the workflow you described, if needed and if we are ok having the communication going through the `FunctionRunner`, otherwise we could let `FunctionRunners` provide a grpc endpoint back to crossplane for functions too so that Crossplane could directly communicate with the Function. Obviously inlining this kind of functions would be inefficient as we would have to spin up the deployment, but could still be doable in some way.

Reply:
I'm not sure I follow what having separate `FunctionRunner` and `Function` types buys us that is more desirable than @negz's current proposal.  I like the simplicity of the `Function` abstraction @negz has proposed where the details about how its run is not something the platform operator has to worry about.  I feel it keeps the concepts simple all around 😊

Reply:
@jbw976, it only feels less confusing to me, but as we said it’s mainly cosmetics and is not changing the end result. However given that we are proposing to  promote it to Beta, I feel it’s worth discussing about the API and trying to make the UX as smooth as possible.

I think the simplicity of just having Functions is not going to last long, we are almost surely going to have something like a FunctionConfig down the line, to make Functions+config reusable, at least for the “runtime” based functions, such as the starlark example below, where the config is actually used to provide custom logic and it’s going to be really painful to duplicate across multiple Compositions if needed. Which is probably going to overlap with the config to run the Function server, which we’ll most probably need too down the line in some form.

So, instead of stretching the concept of Function all over the places and having Functions that are actually servers and “runtime” Functions that run custom user-provided functions and then complex Functions that are actually single purpose servers built by users. I feel more natural and clear as an API to have FunctionRunners that run Functions, this way “runtime” functions are actually FunctionRunners and function+config becomes just a Function, and complex user-provided functions just become Functions run by a specific FunctionRunner spinning up deployments.

Reply:
I prefer to avoid introducing a new `FunctionRunner` concept to our users and keep it as implementation details as long as it makes sense.

I agree with @phisco's concerns around the need for reusable config's for functions, however, would like to see it as `Function` (as package) + `FunctionConfig` (as CRD), which is consistent with what we have today with `Provider` (as package) + `ProviderConfig` (as CRD). (This could also help with the passing credential problem in #3718 /cc @ezgidemirel).

Reply:
After some discussion with @turkenh I've renamed `config` to `input`. This better communicates what this type is for. It doesn't configure the `Function` type in any global way - it just specifies optional input passed with the Function call.

Reply:
Wouldn’t this mean that a function would have two types of inputs? The config and the actual input coming from the P&T pipeline or TM the previous function?

Reply:
> Wouldn’t this mean that a function would have two types of inputs? The config and the actual input coming from the P&T pipeline or the previous function?

Yes - that's what I was trying to communicate before. The input to a "function call" is like `function(xr, composed_resources, input)`.

The mental model of a call being more like `runner(input)(XR, composed_resources)` does conceptually make sense to me, especially given that we think some "inputs" will be repeated a lot - i.e. N Compositions might all call the same function with the same input (fka config).

I still prefer to not expose "runner" as a distinct concept just because:

* There's fewer concepts to think about overall. You just install ("define") a Function, then call it (from a Composition).
* Similarly, there's fewer API types to worry about. In many cases you just install a `Function` (often as a dependency of your Configuration) and call it from your Composition. You don't need to install a `FunctionRunner`, then define a `Function`, then call that from your Composition.

Reply:
I still think most people are going to use “runtime” based functions, at least that’s what I would do. So the two options are actually Function (+ PackageRuntimeConfig) + FunctionConfig or FunctionRunner (+ PackageRuntimeConfig) + Function. I still think the latter would be clearer to me, and it’s actually more similar to the original proposal, just with less focus on OCI images. However as we said it’s mostly cosmetic, so I’m ok marking this thread as resolved given that it looks like there is definitely consensus around the proposed solution.